Functions/Quantification/xASL_quant_External.m

```diff
@@ -0,0 +1,706 @@
+	case 'vaby'
+		ExternalOptions = [ExternalOptions ' --save-var'];
+		ExternalOptions = [ExternalOptions ' --save-residuals'];
+		ExternalOptions = [ExternalOptions ' --allow-bad-voxels'];
+		ExternalOptions = [ExternalOptions ' --save-model-fit'];
+		ExternalOptions = [ExternalOptions ' --max-iterations=100'];
+end
+
+switch (lower(strQuantificationType))
+	case 'fabber'
+		% Basic tissue parameters
+		fprintf(FIDoptionFile, '--t1b=%f\n', x.Q.BloodT1/1000);
+		fprintf(FIDoptionFile, '--t1=%f\n', x.Q.TissueT1/1000);
+
+		% T2-times needed for multi-TE quantification
+		fprintf(FIDoptionFile, '--t2b=%f\n', x.Q.T2art/1000);
+		fprintf(FIDoptionFile, '--t2=%f\n', x.Q.T2/1000);
+	case 'basil'
+		% Basic tissue parameters
+		fprintf(FIDoptionFile, '--t1b=%f\n', x.Q.BloodT1/1000);
+		fprintf(FIDoptionFile, '--t1=%f\n', x.Q.TissueT1/1000);
+
+	case 'vaby'
+		% Basic tissue parameters
+		ExternalOptions = [ExternalOptions, sprintf(' --t1b=%f', x.Q.BloodT1/1000)];
+		ExternalOptions = [ExternalOptions, sprintf(' --t1=%f', x.Q.TissueT1/1000)];
+
+		% T2-times needed for multi-TE quantification
+		ExternalOptions = [ExternalOptions, sprintf(' --t2b=%f', x.Q.T2art/1000)];
+		ExternalOptions = [ExternalOptions, sprintf(' --t2=%f', x.Q.T2/1000)];
+end
+
+%% 3. Basic acquisition parameters
+switch lower(x.Q.LabelingType)
+	% PASL quantification
+	case 'pasl'
+		% PASL model is assumed by default and does not need to be specified in the config file
+		fprintf('BASIL: PASL model\n');
+		
+		if ~isempty(regexpi(strQuantificationType, '(FABBER|VABY)'))
+			error('PASL is implemented for BASIL only and not for FABBER/VABY');
+		end
+
+		% For PASL, there can be only a single LabelingDuration, so unique PLD+LabDur combinations are uniquely based on PLDs
+		TIs = jsonPWI4D.Q.Initial_PLD'/1000;
+
+		% Print all the TIs
+		if bQuantifyMultiPLD	
+			for iTI = 1:length(TIs)
+				fprintf(FIDoptionFile, '--ti%d=%.2f\n', iTI, TIs(iTI));
+			end
+		else
+			fprintf(FIDoptionFile, '--ti=%.2f\n', TIs);
+		end
+
+		% Either print bolus duration or unspecify it
+		if isfield(jsonPWI4D.Q, 'LabelingDuration') && ~isempty(jsonPWI4D.Q.LabelingDuration) 
+			if length(unique(jsonPWI4D.Q.LabelingDuration))>1
+				warning('PASL multi-PLD currently supports only a single Labeling Duration');
+			end
+			fprintf(FIDoptionFile, '--tau=%.2f\n', jsonPWI4D.Q.LabelingDuration(1)/1000);
+		else
+			% Bolus duration not know. If multi-TI, then try to infer it
+			if length(TIs) > 1
+				fprintf(FIDoptionFile, '--infertau\n');
+				fprintf('BASIL: Infer bolus duration component\n')
+			end
+		end
+
+	% CASL and PCASL quantification
+	case {'casl','pcasl'}
+		% Prepare unique PLDs+LabDur combinations
+		
+		% First create a labeling duration vector of the same length
+		LabDurs = jsonPWI4D.Q.LabelingDuration/1000;
+		PLDs = jsonPWI4D.Q.Initial_PLD/1000;
+
+		switch (lower(strQuantificationType))
+			case 'fabber'
+				% For FABBER and multi-TE, we have to group TEs
```

Comment:
Good points. 
It would be simple to implement multi-TE PASL, but we don't have such an example, so I prefer not to do it, because I'm not able to test it. So I added a warning about this in 385a0d4d

Of course if we do internal ExploreASL quantification. Or if multi-TE is removed from PWI4D.json, then we won't notice it here and everything will run fine. So no edit need for that.

Reply:
OK then only my question remains if we proceed to multi-TE quantification and the user has only 3 TEs or something, do you want to issue a warning that this is perhaps not robust, or not? What is the minimum number of TEs required for a decent fit?

We could do the same for PLD of course.

But you can also ignore this point and resolve

Reply:
Good point. 3 is a bare minimum, less than 3 cannot work. 

For less than 3, I issue an error.
For the rest, I would not care - a bit of user responsibility. These data are not so easy to come around, so people won't run it on 3TEs by a mistake ;)